stats/registry.go

```diff
@@ -0,0 +1,50 @@
+package stats
+
+import (
+	"fmt"
+	"sync"
+)
+
+var errFmtMetricExists = "fatal: metric %q already exists"
+
+// Registry tracks metrics and reporters
+type Registry struct {
+	sync.Mutex
+	// here we use just the metric name as key. it does not include any prefix
+	// this means that technically we're more strict here then needed around naming conflicts:
+	// it's possible to have metrics with the same name, but a different prefix.
+	// we still complain about that
+	metrics map[string]GraphiteMetric
+}
+
+func NewRegistry() *Registry {
+	return &Registry{
+		metrics: make(map[string]GraphiteMetric),
+	}
+}
+
+func (r *Registry) add(name string, metric GraphiteMetric) GraphiteMetric {
+	r.Lock()
+	if _, ok := r.metrics[name]; ok {
+		panic(fmt.Sprintf(errFmtMetricExists, name))
```

Comment:
I really dont think this is an un-recoverable error.  We should return an error instead.

Perhaps it would be better to use an addOrGet() method that either returns an existing graphiteMetric with the specified name or creates the new one.